### ISSUE TITLE
Performance issues!?

### ISSUE THREAD

**Issue by gustafnilklint** [nuclearace/Socket.IO-Client-Swift#17]
Hi there!
First of all tanks for a great framework that is Easy to setup and get started.
We recently used this framework in a lab to learn socket IO. A multiplayer version of the classic game Pong.
This exposes possible weakness of this framework.
After the two players has connected and started the game the server will emit the current "state" of the game 30 times per second.
This seems to be problematic since the socket seems to be blocking the main thread (ui thread).
The result is that the game stutters.
All form of network communication should be done on a background thread not blocking the ui.
Have you done any stress tests?

**Comment by nuclearace** [comment]
I'll look into it. I thought I was doing all the background stuff on another queue. But I know the handlers get executed on the main queue.

**Comment by nuclearace** [comment]
@gustafnilklint Just to make sure, you are compiling on -O?

**Comment by calzoneman** [comment]
@gustafnilklint I'm more confused about why you need to send 30 socket messages per second to implement pong...

**Comment by gustafnilklint** [comment]
@nuclearace what do you mean -O, I have tried running it both in debug mode and in release if you mean otimization?
@calzoneman It's not that we "need to" send 30 messages per second.. I can only assume that it was a convenient implementation decision made by the guy that implemented the server for our lab.
It can be mentioned that we split up into teams and made both an Android client, windows phone and a web version aswell. The other clients could handle the pressure just fine without stutter and lag.
All using the same server "spamming" it's clients with 30 updates a second.
At first i thought that it was the server beeing choked with ~20 simulatanious users... But it's just as bad when I run the server on localhost and use the iPhone Simulator.

**Comment by nuclearace** [comment]
@gustafnilklint can you provide a sample message

**Comment by gustafnilklint** [comment]
[players: {
    player1 =     {
        name = web;
        score = 1;
    };
    player2 =     {
        name = gus;
        score = 0;
    };
}, ball: {
    radius = 5;
    x = 239;
    "x_speed" = 1;
    y = 144;
    "y_speed" = "-4";
}, remotePlayerPaddle: {
    height = 10;
    width = 50;
    x = 87;
    y = 10;
}, bounds: {
    height = 600;
    width = 400;
}, playerPaddle: {
    height = 10;
    width = 50;
    x = 175;
    y = 580;
}]

**Comment by nuclearace** [comment]
Erm, from in the form of JSON

**Comment by Xaekai** [comment]
I'm just a bit curious why you would need to have a resolution of 30 updates per second for things like player names, field size (bounds) , and score. Perhaps even the players paddle size.

Seems like those could have their own listeners. That would significantly reduce your packet footprint.

**Comment by gustafnilklint** [comment]
@Xaekai you are absolutely correct, there are plenty of room for improvement and optimisation in the game.
I'm not aiming at rewriting the server here. This could be a server spamming the client with decimal-digits of Pi. But the issue at hand is that the socket implementation is not handling the heavy load very well..

**Comment by nuclearace** [comment]
Keep in mind this is with the #16 branch.

Receiving:

-ONone + polling/WebSockets:
```
Stress lasted: 2.22495722770691. 
Processing 182 events took: 0.171332836151123 seconds from SocketEngine#parseEngineMessage to handler
With an average per event of 0.000941389209621555
```

-O + polling/WebSockets
```
Stress lasted: 2.23535513877869
processing 180 events took: 0.0548372268676758 seconds from SocketEngine#parseEngineMessage to handler
With an average per event of 0.000304651260375977
```

-ONone + forced polling
```
Stress lasted: 2.22789192199707
processing 185 events took: 0.116979837417603 seconds from SocketEngine#parseEngineMessage to handler
With an average per event of 0.000632323445500554
```

-O + forced polling
```
Stress lasted: 2.22730898857117
processing 186 events took: 0.0411303043365479 seconds from SocketEngine#parseEngineMessage to handler
With an average per event of 0.000221130668476064
```

**Comment by naartjie** [comment]
@gustafnilklint is it possible that your code in the handlers is causing the lags? As @nuclearace mentioned, those do get invoked on the main queue. 

Just thinking out loud here, but if you are doing something which is updating the view in a suboptimal way, this could cause the 'stutter' no? Do you maybe want to post your handler code here?

**Comment by gustafnilklint** [comment]
Sorry!!
I closed this issue as performance is not an issue (with this socket implementation).
I accidently posted this issue int the wrong github project. It was intended for the https://github.com/MegaBits/SIOSocket project which is the one I used and experienced the performance issues with.
I recently (yesterday) swapped the socketimplementation in the game client to this one you guys are providing, and it runs smooth.
I'm terribly sorry about these false acusations...
